### PR TITLE
Bump dependencies - Sep 2024

### PR DESCRIPTION
**Pull request by timotheeguerin**
Notable: 
- Typescript 5.6: iterator functions  and more
- typescript-eslint 8: new rules, breaking change to some other, found a few bugs/test not actually testing
- eslint 9: new rules
- removed deprecation/deprecation plugin as typescript eslint has rule for it now(and updated) many places where it wasn't reporting correctly

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 import vitest from "eslint-plugin-vitest";
-import { dirname } from "path";
 import tsEslint from "typescript-eslint";
-import { fileURLToPath } from "url";
 
 /** Config that will apply to all files */
 const allFilesConfig = tsEslint.config({
+  linterOptions: {
+    // This is problematic for no-deprecated rule which needs dependencies to be built to generate the d.ts files and know that it is deprecated
+    // reportUnusedDisableDirectives: "off",
```
Comment (chrisradek) [inline eslint.config.js:12]: Did you work around this? Looks like you're not turning off this flag now.

Comment (chrisradek) [reply eslint.config.js:12]: Edit: Oh I see you remove deprecation plugin later

Comment (timotheeguerin) [reply eslint.config.js:12]: ah yes, using the new projectService from eslint allows it to load the source as long as the tsconfig have `references`

File: packages/http-server-csharp/src/interfaces.ts [hunk 2]
```diff
@@ -101,7 +101,7 @@ export class NumericValue extends CSharpValue {
     this.value = value;
   }
   public emitValue(scope?: Scope<string> | undefined): string {
-    return `${this.value}` ?? "0";
+    return `${this.value}`;
```
Comment (chrisradek) [inline packages/http-server-csharp/src/interfaces.ts:104]: This matches current state - wonder if this was meant to be `${this.value ?? 0}`

File: packages/versioning/test/versioning.test.ts [hunk 3]
```diff
@@ -1037,13 +1037,7 @@ describe("versioning: logic", () => {
         `
       )) as { MyService: Namespace };
 
-      const [v1, v2] = runProjections(runner.program, MyService);
-      const w1 = v1.projectedTypes.get(MyService) as Namespace;
-      const w2 = v2.projectedTypes.get(MyService) as Namespace;
-      w1.models.get("Widget")?.properties.size === 1;
-      w1.operations.get("create")?.parameters.properties.size === 1;
-      w2.models.get("Widget")?.properties.size === 2;
-      w2.operations.get("create")?.parameters.properties.size === 2;
```
Comment (chrisradek) [inline packages/versioning/test/versioning.test.ts:1046]: Why were these removed?

Comment (timotheeguerin) [reply packages/versioning/test/versioning.test.ts:1046]: those weren't actually running assertion and tbh after looking at it not sure they had much more value so I just kept the current state of just running the projection and making sure it doesn't carash

**Review by chrisradek — APPROVED** [review]
